### PR TITLE
Implement Function type for semtypes 

### LINKED ISSUE #32337: Implement Function type
**Description:**

**Describe your task(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by ushirask**
Fixes #32337

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: semtypes/src/main/java/io/ballerina/semtype/Conjunction.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Represents the Conjunction record type.
+ *
+ * @since 2.0.0
+ */
+public class Conjunction {
+    public Atom atom;
+    public Conjunction next;
+
+    public Conjunction(Atom atom, Conjunction next) {
```
Comment (rdhananjaya) [inline semtypes/src/main/java/io/ballerina/semtype/Conjunction.java:29]: Shall we make this private

File: semtypes/src/main/java/io/ballerina/semtype/Env.java [hunk 2]
```diff
@@ -42,5 +42,17 @@ public Env() {
         this.recFunctionAtoms = new ArrayList<>();
     }
 
+    public synchronized RecAtom recFunctionAtom() {
+        int result = this.recFunctionAtoms.size();
+        this.recFunctionAtoms.add(null);
```
Comment (rdhananjaya) [inline semtypes/src/main/java/io/ballerina/semtype/Env.java:47]: shall we add a comment here saying we are adding a `()`

File: semtypes/src/main/java/io/ballerina/semtype/FunctionAtomicType.java [hunk 3]
```diff
@@ -22,7 +22,13 @@
  *
  * @since 2.0.0
  */
-public class FunctionAtomicType {
-    SemType paramType;
-    SemType retType;
+public class FunctionAtomicType implements AtomicType {
+    public SemType paramType;
+    public SemType retType;
+
```
Comment (rdhananjaya) [inline semtypes/src/main/java/io/ballerina/semtype/FunctionAtomicType.java:28]: Extra line

File: semtypes/src/main/java/io/ballerina/semtype/RecAtom.java [hunk 4]
```diff
@@ -23,7 +23,7 @@
  * @since 2.0.0
  */
 public class RecAtom implements Atom {
-    int index;
+    public int index;
```
Comment (rdhananjaya) [inline semtypes/src/main/java/io/ballerina/semtype/RecAtom.java:26]: let's make this final too.

File: semtypes/src/main/java/io/ballerina/semtype/Conjunction.java (hunk 1 above)
Comment (ushirask) [reply semtypes/src/main/java/io/ballerina/semtype/Conjunction.java:29]: just the constructor or the fields as well? if latter we'll have to create field access methods as well

File: semtypes/src/main/java/io/ballerina/semtype/Env.java (hunk 2 above)
Comment (manuranga) [reply semtypes/src/main/java/io/ballerina/semtype/Env.java:48]: Java is prone to get null pointer exceptions, so maybe `null` is not the best choice. On the other hand it is closer to the nballerina source.

Another option is to create a consent like below and use that instead. This is similar to what java Enums get desugared into.
```java
public static final FunctionAtomicType NIL_FUNCTION = new FunctionAtomicType();
```
@jclark any preference ?

Comment (rdhananjaya) [reply semtypes/src/main/java/io/ballerina/semtype/Env.java:48]: I'm going to merge this for now, @ushirask please create an issue if we get any change requests

**Pull request merged**